tool/tsh/tsh.go

```diff
@@ -988,6 +997,91 @@ func refuseArgs(command string, args []string) {
 	}
 }
 
+// rawIdentity encodes the basic components of an identity file.
+type rawIdentity struct {
+	PrivateKey []byte
+	Certs      struct {
+		SSH []byte
+		TLS []byte
+	}
+	CACerts struct {
+		SSH [][]byte
+		TLS [][]byte
+	}
+}
+
+// decodeIdentity attempts to break up the contents of an identity file
+// into its respective components.
+func decodeIdentity(r io.Reader) (*rawIdentity, error) {
+	scanner := bufio.NewScanner(r)
+	var ident rawIdentity
+	// Subslice of scanner's buffer pointing to current line
+	// with leading and trailing whitespace trimmed.
+	var line []byte
+	// Attempt to scan to the next line.
+	scanln := func() bool {
+		if !scanner.Scan() {
+			line = nil
+			return false
+		}
+		line = bytes.TrimSpace(scanner.Bytes())
+		return true
+	}
+	// Check if the current line starts with prefix `p`.
+	peekln := func(p string) bool {
+		return bytes.HasPrefix(line, []byte(p))
+	}
+	// Get an "owned" copy of the current line.
+	cloneln := func() []byte {
+		ln := make([]byte, len(line))
+		copy(ln, line)
+		return ln
+	}
+	// Scan through all lines of identity file.  Lines with a known prefix
+	// are copied out of the scanner's buffer.  All others are ignored.
+	for scanln() {
+		switch {
+		case peekln("ssh"):
+			ident.Certs.SSH = cloneln()
+		case peekln("@cert-authority"):
+			ident.CACerts.SSH = append(ident.CACerts.SSH, cloneln())
+		case peekln("-----BEGIN"):
+			// Current line marks the beginning of a PEM block.  Consume all
+			// lines until a corresponding END is found.
+			var pemBlock []byte
+			for {
```

Comment:
mostly curious, is there a chance for an endless loop here?

Reply:
Worst-case, the loop should always terminates on EOF, because we only continue iteration if the scanner is able to make progress.

Reply:
Never hurts to ask!